Expand MSBuildSdkResolver

surayya-MS opened the pull request:
Fixes https://github.com/dotnet/msbuild/issues/11086

### Context
Details - [Here's the flow for sdk resolution in msbuild repo](https://github.com/dotnet/msbuild/blob/6ca999210ce0e6541d92532874b732c4fceee8de/src/Build/BackEnd/Components/SdkResolution/SdkResolverService.cs#L113-L138)
`MSBuildSdkResolver` is in resolver list only in case of `msbuild.exe`. 
`MSBuildSdkResolver` is not present in case of `dotnet build`. Hence, `Muxer` is not available in `msbuild.exe` context as the corresponding issue description suggests.

### Changes
1. Added `DOTNET_HOST_PATH` to the `SdkResult.PropertiesToAdd` with value of the path of `dotnet.exe` similar to [dotnet CLI does before triggering builds](https://github.com/dotnet/sdk/blob/bac9b27723ae9d3dc4f6fb8f47d9d2da1a0a0a0c/src/Cli/Microsoft.DotNet.Cli.Utils/MSBuildForwardingAppWithoutLogging.cs#L49)
2. Added `SdkResolverMSBuildTaskHostRuntimeVersion `  to the `SdkResult.PropertiesToAdd` with value of  the runtime version MSBuild uses (from `MSBuild.runtimeconfig.json` file) similar to [`Muxer.SharedFxVersion`](https://github.com/dotnet/sdk/blob/5433bc4745265d7c16611f2b389b780e834eacc0/src/Cli/Microsoft.DotNet.Cli.Utils/Muxer.cs#L14-L21)

### Testing
Manually tested by building sdk repo with `build.cmd`, then copying `Microsoft.DotNet.MSBuildSdkResolver.dll` to msbuild, then running `msbuild.exe` on a console app. Result - `DOTNET_HOST_PATH` and `SdkResolverMSBuildTaskHostRuntimeVersion ` are present in the binlog as a property in Evaluation phase.
Added check for `DOTNET_HOST_PATH` and `SdkResolverMSBuildTaskHostRuntimeVersion ` keys in unit tests.

surayya-MS committed:
add DOTNET_HOST_PATH to MSBuildSdkResolver

baronfel commented on src/Resolvers/Microsoft.DotNet.MSBuildSdkResolver/MSBuildSdkResolver.cs:
```diff
@@ -189,6 +189,13 @@ private sealed class CachedState
                         minimumVSDefinedSDKVersion);
                 }
 
+                string dotnetExe = Path.Combine(dotnetRoot, "dotnet.exe");
```
Do we have `Muxer.cs` available from `Microsoft.DotNet.Cli.Utils` at this point? We should use `new Muxer().MuxerPath` for the path to `dotnet.exe`, and we should use `new Muxer().SharedFxVersion` as the value for the `MSBUILD_NET_TASKHOST_RUNTIME_VERSION` property.

surayya-MS replied:
When I tried to run `dotnet build` to build a console app, the `MSBuildSdkResolver` was not in the list of the resolvers. When I tried to run `msbuild` the `MSBuildSdkResolver` was in the list. I also found this comment in msbuild repo that explains the flow of sdk resolution:
https://github.com/dotnet/msbuild/blob/6ca999210ce0e6541d92532874b732c4fceee8de/src/Build/BackEnd/Components/SdkResolution/SdkResolverService.cs#L113-L138

`MuxerPath` gets the current process's path. In case of `dotnet` it gets the path to `dotnet` that is running.

baronfel replied:
ah, of course - for `dotnet build` there is no MSBuildSdkResolver because the SDK sets up the variables required for the MSBuild DefaultSdkResolver to Just Work.

baronfel commented:
When this is merged, we will want to backport to release/9.0.2xx so that this will ship in a version of VS that Roslyn can use to begin testing. If we don't backport, they would have to wait until we had an insertion of .NET 10 into VS, which is quite a while away.

surayya-MS committed:
fix tests

surayya-MS committed:
Merge branch 'main' into expand-MSBuildSdkResolver

surayya-MS committed:
assert count in test

surayya-MS committed:
use consts to get "dotnet.exe"

surayya-MS committed:
fix tests

surayya-MS committed:
Merge branch 'main' into expand-MSBuildSdkResolver

surayya-MS committed:
remove unnecessary usings

rainersigwald commented on test/Microsoft.DotNet.MSBuildSdkResolver.Tests/GivenAnMSBuildSdkResolver.cs:
```diff
@@ -274,7 +278,9 @@ public void ItReturnsHighestSdkAvailableThatIsCompatibleWithMSBuildWhenVersionIn
             result.AdditionalPaths.Should().BeNull();
-            result.PropertiesToAdd.Count.Should().Be(2);
+            result.PropertiesToAdd.Count.Should().Be(4);
+            result.PropertiesToAdd.Should().ContainKey(_dotnetHost);
+            result.PropertiesToAdd.Should().ContainKey(_msbuildNetTaskHostRuntimeVersion);
             result.PropertiesToAdd.ContainsKey("SdkResolverHonoredGlobalJson");
             result.PropertiesToAdd.ContainsKey("SdkResolverGlobalJsonPath");
```
What's the difference between `.Should().ContainKey()` and `.ContainsKey()`?

surayya-MS replied:
Interesting. `ContainsKey` just returns bool. And then there is no assertion. It was probably meant to be `Shgould().ContainKey()` instead. I will fix this

rainersigwald commented on src/Resolvers/Microsoft.DotNet.MSBuildSdkResolver/MSBuildSdkResolver.cs:
```diff
@@ -190,6 +191,22 @@ private sealed class CachedState
                 }
 
+                string dotnetExe = Path.Combine(dotnetRoot, Constants.DotNet + Constants.ExeSuffix);
+                if (File.Exists(dotnetExe))
```
This exists check could lead to a silent failure. Is that the right behavior?

surayya-MS replied:
not sure. @baronfel should we throw an exception if "dotnet.exe" file does not exist? or we just don't set `DOTNET_HOST_PATH`?
If we decide to throw should this be under ChangeWave? Not sure how to achieve that in sdk repo. Probably to have 2 versions of this resolver

YuliiaKovalova replied:
If we switch to [/EnvironmentProvider.cs](https://github.com/dotnet/sdk/blob/8bd19a2390a6bba4aa80d1ac3b6c5385527cc311/src/Resolvers/Microsoft.DotNet.NativeWrapper/EnvironmentProvider.cs#L52), the info about missed dotnet will be logged in the method

surayya-MS replied:
I decided to add logging in case when `DOTNET_HOST_PATH` or `SdkResolverMSBuildTaskHostRuntimeVersion` could not be set. Feel free to unresolve this conversation, if you decide we should throw

rainersigwald commented on src/Resolvers/Microsoft.DotNet.MSBuildSdkResolver/MSBuildSdkResolver.cs:
```diff
@@ -190,6 +191,22 @@ private sealed class CachedState
+                if (File.Exists(dotnetExe))
+                {
+                    propertiesToAdd ??= new Dictionary<string, string?>();
+                    propertiesToAdd.Add("DOTNET_HOST_PATH", dotnetExe);
```
I feel medium-strongly that `DOTNET_HOST_PATH` shouldn't be the property name, since it's against convention. Maybe
```suggestion
                    propertiesToAdd.Add("SdkResolverDotnetHostPath", dotnetExe);
```
to match the `global.json` reports?

baronfel replied:
One of the reasons we talked about using the DOTNET_HOST_PATH property was to take advantage of all of the existing logic that already relied on DOTNET_HOST_PATH if it was set - if we change the name we'd either need

* an early sdk-level mapping in the props from SdkResolverDotnetHostPath => DOTNET_HOST_PATH, or
* each of the current consumers of DOTNET_HOST_PATH to do this same mapping

surayya-MS replied:
One thing to mention - in "dotnet build" case `DOTNET_HOST_PATH` is also an environment variable. Should we set this env variable here as well or property is sufficient?

surayya-MS replied:
resolving this conversation

rainersigwald commented on src/Resolvers/Microsoft.DotNet.MSBuildSdkResolver/MSBuildSdkResolver.cs:
```diff
@@ -190,6 +191,22 @@ private sealed class CachedState
+                if (runtimeVersion != null)
+                {
+                    propertiesToAdd ??= new Dictionary<string, string?>();
+                    propertiesToAdd.Add("MSBUILD_NET_TASKHOST_RUNTIME_VERSION", runtimeVersion);
```
(same)

baronfel replied:
This one I'm way more flexible on because I expect it to be a value that the MSBuild engine itself would start using once we start implementing the NetCore TaskHost, so we are more fully in the driver's seat here.

surayya-MS replied:
changed to `MSBuildNetTaskHostRuntimeVersion`

surayya-MS replied:
Oops. The convention is to start with "SdkResolver...". Renamed to "SdkResolverMSBuildTaskHostRuntimeVersion" (removed "Net" part because it is too long)

rainersigwald commented on src/Resolvers/Microsoft.DotNet.MSBuildSdkResolver/MSBuildSdkResolver.cs:
```diff
@@ -259,6 +273,31 @@ private sealed class CachedState
+            string runtimeConfigPath = Path.Combine(sdkDirectory, "MSBuild.runtimeconfig.json");
+            if (File.Exists(runtimeConfigPath))
+            {
+                var runtimeConfigJson = JsonNode.Parse(File.ReadAllText(runtimeConfigPath)) as JsonObject;
```
Is there a more efficient way to read these few fields out of the stream than "fully read it into memory then fully deserialize it"? Probably not _super_ important here but this is allocations in MSBuild.exe . . .

surayya-MS replied:
the file `MSBuild.runtimeconfig.json` is very short. Example from the built dotnet artifacts:
```json
{
  "runtimeOptions": {
    "tfm": "net10.0",
    "framework": {
      "name": "Microsoft.NETCore.App",
      "version": "10.0.0-alpha.1.24610.2"
    },
    "configProperties": {
      "System.Runtime.Serialization.EnableUnsafeBinaryFormatterSerialization": false
    }
  }
}
```

Is it a constant thing or the file can be bigger?

baronfel replied:
It _can_ get bigger, but it's going to say on this same order of magnitude. Compare this to the one for .NET 9, for example:

```json
 cat 'C:\Program Files\dotnet\sdk\9.0.101\MSBuild.runtimeconfig.json'
{
  "runtimeOptions": {
    "tfm": "net9.0",
    "framework": {
      "name": "Microsoft.NETCore.App",
      "version": "9.0.0"
    },
    "configProperties": {
      "System.Reflection.Metadata.MetadataUpdater.IsSupported": false,
      "System.Runtime.Serialization.EnableUnsafeBinaryFormatterSerialization": false
    }
  }
}
```

The main thing that will change across releases is what `configProperties` that will be present - and in general that should be quite small.

surayya-MS replied:
I decided to go the safer route and read json from a stream

surayya-MS committed:
fix the test

surayya-MS committed:
mock getting runtime version for tests

surayya-MS committed:
rename prop to MSBuildNetTaskHostRuntimeVersion

surayya-MS committed:
Merge branch 'main' into expand-MSBuildSdkResolver

YuliiaKovalova commented on src/Resolvers/Microsoft.DotNet.MSBuildSdkResolver/MSBuildSdkResolver.cs:
```diff
@@ -190,6 +194,22 @@ private sealed class CachedState
                         minimumVSDefinedSDKVersion);
                 }
 
+                string dotnetExe = Path.Combine(dotnetRoot, Constants.DotNet + Constants.ExeSuffix);
```
The ExeSuffix does not exist in Linux/OS environments (executable is called `dotnet`)
It looks like you can get the path safely by using existing util class:
https://github.com/dotnet/sdk/blob/8bd19a2390a6bba4aa80d1ac3b6c5385527cc311/src/Resolvers/Microsoft.DotNet.NativeWrapper/EnvironmentProvider.cs#L52

baronfel replied:
Just a note that this resolved is only used by VS, so is framework and Windows only. Concerns like getting the non-exe suffix do not apply.

surayya-MS replied:
thanks!
I checked and `GetDotnetExeDirectory` returns the directory of dotnet executable. `DOTNET_HOST_PATH` should be equal to the path of the executable (it is in `dotnet build` case)

I added a new const "dotnet.exe" and check only that

YuliiaKovalova commented on src/Resolvers/Microsoft.DotNet.MSBuildSdkResolver/MSBuildSdkResolver.cs:
```diff
@@ -259,6 +276,31 @@ private sealed class CachedState
+                string? runtimeVersion = runtimeOptionsFramework?["version"]?.ToString();
+
+                // 2. Check that the runtime version is installed (in shared folder)
+                if (runtimeName != null && runtimeVersion != null)
```
please also check if these fields don't hold empty values

surayya-MS replied:
thanks! done.

surayya-MS committed:
rename to SdkResolverMSBuildTaskHostRuntimeVersion

surayya-MS committed:
small refactoring

surayya-MS committed:
rename variable

surayya-MS committed:
check string is null or empty;
check only "dotnet.exe";
log if couldn't set DOTNET_HOST_PATH and  SdkResolverMSBuildTaskHostRuntimeVersion

surayya-MS committed:
read json fields in a stream

surayya-MS committed:
add comment

surayya-MS committed:
Merge branch 'main' into expand-MSBuildSdkResolver

surayya-MS committed:
fix message

rainersigwald commented on test/Microsoft.DotNet.MSBuildSdkResolver.Tests/GivenAnMSBuildSdkResolver.cs:
```diff
@@ -200,7 +202,9 @@ public void ItReturnsHighestSdkAvailableThatIsCompatibleWithMSBuild(bool disallo
-            result.PropertiesToAdd.Should().BeNull();
+            result.PropertiesToAdd.Count.Should().Be(2);
+            result.PropertiesToAdd.Should().ContainKey(_dotnetHost);
+            result.PropertiesToAdd.Should().ContainKey(_msbuildTaskHostRuntimeVersion);
```
Validate the `mockRuntimeVersion` value too?

surayya-MS replied:
thanks! done.

surayya-MS committed:
check the value of SdkResolverMSBuildTaskHostRuntimeVersion in tests;
rename const variables

surayya-MS commented on src/Resolvers/Microsoft.DotNet.MSBuildSdkResolver/MSBuildSdkResolver.cs:
```diff
@@ -28,23 +29,28 @@ public sealed class DotNetMSBuildSdkResolver : SdkResolver
         private readonly NETCoreSdkResolver _netCoreSdkResolver;
 
+        private const string _dotnetHost = "DOTNET_HOST_PATH";
+        private const string _msbuildTaskHostRuntimeVersion = "SdkResolverMSBuildTaskHostRuntimeVersion";
```
thanks! done.

surayya-MS committed:
rename const

surayya-MS commented:
/backport to release/9.0.2xx
